### ISSUE TITLE
Matmul - improve pass rate and performance with default settings

### ISSUE THREAD

**Issue by cglagovichTT** [tenstorrent/tt-metal#29716]
## Context
TT-DiT is implementing many image and video gen models. Performance of these workloads is critical. In these workloads, matmul accounts for 20-25% of latency on device.
Today our typical usage of matmul is the following:
- interleaved inputs/output
- BF16, tilized tensors
- HiFi2 fidelity, FP32 accumulate, packer accumulation enabled

We use packer accumulation because in practice it results in much better correctness. I typically see 10x worse MSE with packer accumulation disabled. 
To invoke matmul, our standard interface is
```
ttnn.linear(x, weight, bias=bias, core_grid=core_grid, compute_kernel_config=compute_kernel_config)
```
We pass `core_grid` because that leads linear to fuse the bias addition into matmul. Fused bias (and activation) are necessary for performance. 
We don't generate or pass a program config because
- TT-DiT should be a generalized library without complex input and model-specific program configs baked in
- it is difficult to find valid and performant program configs

With our current setup, I am measuring that matmul utilization is in the range of 10.5% to 47.2%, mean=32.3%, median=38.2%. Inputs are large (typically each dim >=2048), so I'd expect we can very much improve utilization. My target is 60%. 

## Request
I have a sweep test on branch for the matmul shapes I care about. In this sweep test,
- 1 case fails L1 out-of-memory
- all cases achieve below the target of 60% utilization

For the given repro, please ensure that there are no L1 OOM issues, and that all MMs achieve > 60% MM util.

## Repro
branch [cglagovich/29716](https://github.com/tenstorrent/tt-metal/tree/cglagovich/29716)

### accuracy test
```
pytest models/experimental/tt_dit/tests/unit/performance/test_linear.py::test_linear
```
output: all tests pass except for `test_linear[mochi_ff1_2x4sp1tp0]`

### performance test
```
pytest models/experimental/tt_dit/tests/unit/performance/test_linear.py::test_summarize_linear_performance
```

Output on N300 (8x8 coregrid):
```
| M, K, N | measured perf (ms) | math util (%) | expected perf at 60% util (ms) |
|---|---:|---:|---:|
| (11264, 3072, 4608) | 6.230 | 39.1 | 4.055 |
| (11264, 3072, 8192) | - | - | - |
| (11264, 4096, 3072) | 4.579 | 47.2 | 3.604 |
| (5632, 3072, 2304) | 1.460 | 41.7 | 1.014 |
| (5632, 3072, 4096) | 2.531 | 42.7 | 1.802 |
| (5632, 2048, 3072) | 1.410 | 38.3 | 0.901 |
| (37888, 5120, 1280) | 31.100 | 12.4 | 6.415 |
| (37888, 5120, 3456) | 76.216 | 13.6 | 17.320 |
| (37888, 3456, 5120) | 26.077 | 39.2 | 17.050 |
| (9472, 5120, 1280) | 9.041 | 10.5 | 1.579 |
| (9472, 5120, 3456) | 13.110 | 19.5 | 4.262 |
| (9472, 3456, 5120) | 12.265 | 20.9 | 4.262 |
| (9472, 5120, 2560) | 9.359 | 20.2 | 3.157 |
| (9472, 5120, 6912) | 17.386 | 29.4 | 8.525 |
| (9472, 6912, 5120) | 22.865 | 22.4 | 8.525 |
```

Output on 6U (8x9 coregrid):
```
| M, K, N | measured perf (ms) | math util (%) | expected perf at 60% util (ms) |
|---|---:|---:|---:|
| (11264, 3072, 4608) | 5.665 | 38.2 | 3.604 |
| (11264, 3072, 8192) | 10.148 | 37.9 | 6.408 |
| (11264, 4096, 3072) | 4.738 | 40.6 | 3.204 |
| (5632, 3072, 2304) | 1.361 | 39.7 | 0.901 |
| (5632, 3072, 4096) | 2.601 | 37.0 | 1.602 |
| (5632, 2048, 3072) | 1.445 | 33.3 | 0.801 |
| (37888, 5120, 1280) | 25.058 | 13.8 | 5.773 |
| (37888, 5120, 3456) | 73.948 | 12.6 | 15.588 |
| (37888, 3456, 5120) | 23.160 | 39.3 | 15.155 |
| (9472, 5120, 1280) | 2.453 | 34.3 | 1.403 |
| (9472, 5120, 3456) | 5.437 | 41.8 | 3.789 |
| (9472, 3456, 5120) | 5.821 | 39.1 | 3.789 |
| (9472, 5120, 2560) | 3.940 | 42.7 | 2.807 |
| (9472, 5120, 6912) | 9.880 | 46.0 | 7.578 |
| (9472, 6912, 5120) | 10.197 | 44.6 | 7.578 |
```

**Comment by bbradelTT** [comment]
@cglagovichTT does the perf model on your branch take into account the bias calculation?

**Comment by cglagovichTT** [comment]
@bbradelTT no, I'm assuming that bias addition is insignificant, which I believe is fair.

**Comment by bbradelTT** [comment]
Note: may need to factor in TT_MM_THROTTLE_LEVEL

**Comment by bbradelTT** [comment]
Should test with and without core_grid

**Comment by cglagovichTT** [comment]
I am okay setting this issue to P2 since we have a workaround for DiT use cases - `minimal_matmul` hits our performance targets.
Let me know if you agree with the change in priority.

**Comment by bbradelTT** [comment]
This is the next project for @rmillerTT. We can leave it at P1.

**Comment by bbradelTT** [comment]
Extra shapes mentioned in description:
https://github.com/tenstorrent/tt-metal/pull/32295#discussion_r2595222353

nanogpt example:
Matmul manual transpose info:
Config: MatmulMultiCoreReuseProgramConfig
Shape A: Shape([64, 6, 256, 64]), Shape B: Shape([64, 6, 256, 64])
Transpose A: false, Transpose B: true
Dtype A: DataType::BFLOAT16, Dtype B: DataType::BFLOAT16
gpt2 example:
Matmul manual transpose info:
Config: MatmulMultiCoreReuseProgramConfig
Shape A: Shape([4, 12, 1024, 1024]), Shape B: Shape([4, 12, 1024, 64])
Transpose A: true, Transpose B: false
Dtype A: DataType::BFLOAT16, Dtype B: DataType::BFLOAT16
tinyllama example:
Matmul manual transpose info:
Config: MatmulMultiCoreReuseProgramConfig
Shape A: Shape([1, 32, 2048, 2048]), Shape B: Shape([1, 32, 2048, 64])
Transpose A: true, Transpose B: false
Dtype A: DataType::BFLOAT16, Dtype B: DataType::BFLOAT16

Example pytest (has extra info, may need to be adjusted):
```
import pytest
import torch

import ttnn
from tests.ttnn.utils_for_testing import assert_with_pcc

from models.common.utility_functions import comp_ulp, comp_pcc
@pytest.mark.parametrize("params", [
    ([64,6,256,64], [64,6,256,64], False, True),
    ([4,12,1024,1024], [4,12,1024,64], True, False),
    ([1,32,2048,2048], [1,32,2048,64], True, False),
    #([800,151936], [800,384], True, False),
    ])
@pytest.mark.parametrize("fidelity", [
    #ttnn.MathFidelity.LoFi,
    #ttnn.MathFidelity.HiFi2,
    ttnn.MathFidelity.HiFi4,
    ])
def test_matmul(device, params, fidelity):
    torch.manual_seed(0)
    input_shape, input_shape2, transpose_a, transpose_b = params
    torch_input_tensor = torch.rand(input_shape, dtype=torch.bfloat16)
    torch_input_tensor2 = torch.rand(input_shape2, dtype=torch.bfloat16)

    input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat16,
            layout=ttnn.TILE_LAYOUT,
            device=device)
    input_tensor2 = ttnn.from_torch(torch_input_tensor2, dtype=ttnn.bfloat16,
            layout=ttnn.TILE_LAYOUT,
            device=device)
    if transpose_a:
        torch_input_tensor = torch.transpose(torch_input_tensor, -1, -2)
    if transpose_b:
        torch_input_tensor2 = torch.transpose(torch_input_tensor2, -1, -2)
    torch_output_tensor = torch.matmul(torch_input_tensor, torch_input_tensor2)
    core_grid = core_grid=ttnn.CoreGrid(y=8,x=7)
    #core_grid = core_grid=ttnn.CoreGrid(y=7,x=8)
    #core_grid = None
    compute_kernel_config = ttnn.init_device_compute_kernel_config(
            device.arch(),
            math_fidelity=fidelity,
            math_approx_mode=False,
            fp32_dest_acc_en=True,
            packer_l1_acc=True,
        )
    output_tensor = ttnn.matmul(input_tensor, input_tensor2,
            transpose_a=transpose_a, transpose_b=transpose_b,
            core_grid=core_grid,
            #compute_kernel_config=compute_kernel_config,
            )
    output_tensor = ttnn.to_torch(output_tensor)
    assert_with_pcc(torch_output_tensor, output_tensor, 0.95)
```
